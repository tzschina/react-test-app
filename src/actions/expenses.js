import { v4 as uuid } from 'uuid';

export const addExpense = ({ desc = "", note = "", amount = 0, createdAt = 0 } = {}) => (
    {
        type: "ADD_EXPENSE",
        expense: {
            id: uuid(),
            desc,
            amount,
            createdAt,
            note
        }
    }
);

export const removeExpense = ({ id }) => (
    {
        type: "REMOVE_EXPENSE",
        id
    }
);

export const editExpense = ({ id, updates }) => (
    {
        type: "EDIT_EXPENSE",
        id,
        updates
    }
);

const setTextFilter = (text = '') => (
    {
        type: "SET_TEXT_FILTER",
        text
    }
);