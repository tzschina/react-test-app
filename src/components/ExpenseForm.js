import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

class ExpenseForm extends React.Component {

    constructor(props) {
        super(props);

        this.state = {
            desc: props.expense ? props.expense.desc : '',
            amount: props.expense ? props.expense.amount :  '',
            note: props.expense ? props.expense.note : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            createdAtCalendarFocus: false,
            error: undefined
        }
        this.handleDescriptionChange = this.handleDescriptionChange.bind(this);
        this.handleAmountChange = this.handleAmountChange.bind(this);
        this.handleNoteChange = this.handleNoteChange.bind(this);
        this.handleDateChange = this.handleDateChange.bind(this);
        this.handleCalendarFocusChange = this.handleCalendarFocusChange.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
    }

    handleDescriptionChange(event) {
        const desc = event.target.value;
        this.setState(() => ({ desc }));
    };

    handleAmountChange(event) {
        const amount = event.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        }
    };

    handleNoteChange(event) {
        const note = event.target.value;
        this.setState(() => ({ note }));
    };

    handleDateChange(createdAt) {
        if (createdAt) {
            this.setState(() => ( { createdAt } ));
        }
    }

    handleCalendarFocusChange({ focused }) {
        this.setState(() => ({ createdAtCalendarFocus: focused }));
    }

    handleSubmit(event) {
        event.preventDefault();
        if(!this.state.desc || !this.state.amount) {
            this.setState(() => ({error: 'Description and amount must be entered'}));
        } else {
            this.setState(() => ({error: undefined}));
            this.props.handleExpense(
                {
                    amount: parseFloat(this.state.amount),
                    desc: this.state.desc,
                    note: this.state.note,
                    createdAt: this.state.createdAt.valueOf()
                }
            );
        }
    };

    render() {
        return (
            <div>
                <form onSubmit={this.handleSubmit}>
                    { this.state.error && <div>{this.state.error}</div>}
                    <input type="text" placeholder="Description" autoFocus value={this.state.desc} onChange={this.handleDescriptionChange} />
                    <input type="text" placeholder="Amount" value={this.state.amount} onChange={this.handleAmountChange} />
                    <textarea placeholder="Add note" value={this.state.note} onChange={this.handleNoteChange} />
                    <SingleDatePicker
                        isOutsideRange={() => false}
                        numberOfMonths={1}
                        date={this.state.createdAt}
                        onDateChange={this.handleDateChange}
                        focused={this.state.createdAtCalendarFocus}
                        onFocusChange={this.handleCalendarFocusChange} />
                    <button>{this.props.expense ? 'Update': 'Add'} Expense</button>
                </form>
            </div>
        );
    };
};

export default ExpenseForm;