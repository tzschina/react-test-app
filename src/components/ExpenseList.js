import React from 'react';
import { connect } from 'react-redux';
import getVisibleExpenses from '../selectors/expenses'
import ExpenseListItem from './ExpenseListItem';

export const ExpenseList = (props) => (
    <div className="content-container">
        <div className="list-header">
            <div className="show-for-mobile">Expenses</div>
            <div className="show-for-desktop">Expense</div>
            <div className="show-for-desktop">Amount</div>
        </div>
        {props.expenses.map((expense) => {
            return (
                <ExpenseListItem {...expense} key={expense.id} />
            );
        })}
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: getVisibleExpenses(state.expenses, state.filters)
    }
};


export default connect(mapStateToProps)(ExpenseList);