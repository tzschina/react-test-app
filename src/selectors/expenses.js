import moment from 'moment';

const getVisibleExpenses = (expenses, {text, sortBy, startDate, endDate}) => {

    return expenses.filter((expense) => {
        const createdAtMoment = moment(expense.createdAt)
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day'): true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
        const textMatch = !text || expense.note.toLowerCase().includes(text.toLowerCase()) || expense.desc.toLowerCase().includes(text.toLowerCase());
        return startDateMatch && endDateMatch && textMatch;
    }).sort((expense1, expense2) => {
        if(sortBy === 'date') {
            return expense1.createdAt < expense2.createdAt ? 1: -1;
        } else if (sortBy === 'amount') {
            return expense1.amount < expense2.amount ? 1: -1;
        } else {
            return 0;
        }
    });
};

export default getVisibleExpenses;