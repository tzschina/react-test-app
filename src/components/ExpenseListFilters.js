
import { DateRangePicker } from 'react-dates';
import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';


export class ExpenseListFilters extends React.Component {

    constructor(props) {
        super(props);
        this.state = {
            calendarFocused: null
        };
        this.handleDateChange = this.handleDateChange.bind(this);
        this.handleFocusChange = this.handleFocusChange.bind(this);
        this.handleTextChange = this.handleTextChange.bind(this);
        this.handleSortByChange = this.handleSortByChange.bind(this);
    }

    handleDateChange({ startDate, endDate }) {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    }

    handleFocusChange(calendarFocused) {
        this.setState(() => ({ calendarFocused }));
    }

    handleTextChange(event) {
        this.props.setTextFilter(event.target.value);
    }

    handleSortByChange(event) {
        if (event.target.value === 'date') {
            this.props.sortByDate();
        }
        if (event.target.value === 'amount') {
            this.props.sortByAmount();
        }
    }

    render() {
        return (
            <div className="content-container">
                <div className="input-group">
                    <div className="input-group__item">
                        <input className="text-input" 
                        type='text' 
                        placeholder="Search expenses"
                        value={this.props.filters.text} 
                        onChange={this.handleTextChange} />
                    </div>
                    <div className="input-group__item">
                        <select
                            className="select"
                            value={this.props.filters.sortBy}
                            onChange={this.handleSortByChange}>
                            <option value='date'>Date</option>
                            <option value='amount'>Amount</option>
                        </select></div>
                    <div className="input-group__item">
                        <DateRangePicker
                            startDate={this.props.filters.startDate}
                            endDate={this.props.filters.endDate}
                            onDatesChange={this.handleDateChange}
                            focusedInput={this.state.calendarFocused}
                            onFocusChange={this.handleFocusChange}
                            numberOfMonths={1}
                            isOutsideRange={() => false}
                            showClearDates={true}
                        /></div>
                </div>
            </div>
        )
    };
};




const mapStateToProps = (state) => ({
    filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
    sortByAmount: () => dispatch(sortByAmount()),
    sortByDate: () => dispatch(sortByDate()),
    setTextFilter: (text) => dispatch(setTextFilter(text))
});


export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);