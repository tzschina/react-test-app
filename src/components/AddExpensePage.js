import React from 'react';
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses';
import { connect } from 'react-redux';


export class AddExpensePage extends React.Component {

    constructor(props) {
        super(props);
        this.handleSubmit = this.handleSubmit.bind(this);
    }

    handleSubmit(expense) {
        this.props.startAddExpense(expense);
        this.props.history.push('/');
    }

    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container"><h1>Add Expense</h1></div>
                </div>
                <div className="content-container">
                    <ExpenseForm handleExpense={this.handleSubmit} />
                </div>
            </div>
        );
    }

};


const mapDispatchToProps = (dispatch) => (
    {
        startAddExpense: (expense) => dispatch(startAddExpense(expense))
    }
);

export default connect(undefined, mapDispatchToProps)(AddExpensePage);