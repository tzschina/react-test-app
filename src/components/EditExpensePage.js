import React from 'react';
import ExpenseForm from './ExpenseForm';
import {connect} from 'react-redux';
import {startRemoveExpense, startEditExpense} from '../actions/expenses';

export class EditExpensePage extends React.Component {
    constructor(props) {
        super(props);
        this.handleEdit = this.handleEdit.bind(this);
        this.handleRemove = this.handleRemove.bind(this);
    }

    handleEdit(expense) {
        this.props.startEditExpense({ id: this.props.expense.id, updates: expense});
        this.props.history.push('/');
    }

    handleRemove() {
        this.props.startRemoveExpense(this.props.expense.id);
        this.props.history.push('/');
    }

    render() {
        return (
            <div>
            <ExpenseForm expense={this.props.expense} handleExpense={this.handleEdit}/>
            <button onClick={this.handleRemove}>Remove</button>
            </div>
        )
    };
};


const mapStateToProps = (state, props) => ({
    expense: state.expenses.find((expense) => expense.id === props.match.params.id )
});

const mapDispatchToProps = (dispatch) => ({
    startEditExpense: (update) => dispatch(startEditExpense(update)),
    startRemoveExpense: (id) => dispatch(startRemoveExpense(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);