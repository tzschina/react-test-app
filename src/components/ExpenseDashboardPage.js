import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpenseSummary from './ExpenseSummary';

const ExpenseDahsboardPage = () => (
    <div>
        <ExpenseSummary/>
        <ExpenseListFilters/>
        <ExpenseList/>
    </div>
);


export default ExpenseDahsboardPage;