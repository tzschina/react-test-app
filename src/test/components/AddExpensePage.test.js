import React from 'react';
import {shallow} from 'enzyme';
import expenses from '../fixtures/expenses';
import { AddExpensePage } from '../../components/AddExpensePage';



let startAddExpense, history, wrapper;

beforeEach(() => {
  startAddExpense = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(<AddExpensePage startAddExpense={startAddExpense} history={history}/>);
});

test('should render AddExpensePage correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should handle addExpense', () => {
  wrapper.find('ExpenseForm').prop('handleExpense')(expenses[1]);
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(startAddExpense).toHaveBeenLastCalledWith(expenses[1]);
});
