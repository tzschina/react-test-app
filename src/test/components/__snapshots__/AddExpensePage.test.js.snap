// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AddExpensePage correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1>
        Add Expense
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      handleExpense={[Function]}
    />
  </div>
</div>
`;
